File: 현호/BOJ_1003.js
const fs = require("fs");
const input = fs.readFileSync("/dev/stdin").toString().trim().split("\n");

const testCase = Number(input.shift());
const numbers = input.map(Number);

const solve = () => {
  const dp = Array.from({ length: 41 }, () =>
    Array.from({ length: 2 }, () => -1)
  );
  const answer = [];

  [dp[0][0], dp[0][1]] = [1, 0];
  [dp[1][0], dp[1][1]] = [0, 1];

  for (let i = 0; i < testCase; i += 1) {
    fibonacci(numbers[i], dp);
    answer.push(dp[numbers[i]].join(" "));
  }

  console.log(answer.join("\n"));
};

// 피보나치는 f(n) = f(n - 1) + f(n - 2)
const fibonacci = (n, dp) => {
  if (dp[n][0] === -1) {
    const fibonacci_a =
      dp[n - 1][0] !== -1 ? [dp[n - 1][0], dp[n - 1][1]] : fibonacci(n - 1, dp);
    const fibonacci_b =
      dp[n - 2][0] !== -1 ? [dp[n - 2][0], dp[n - 2][1]] : fibonacci(n - 2, dp);

    [dp[n][0], dp[n][1]] = [
      fibonacci_a[0] + fibonacci_b[0],
      fibonacci_a[1] + fibonacci_b[1],
    ];
  }

  return [dp[n][0], dp[n][1]];
};

solve();

File: 근재/DP/BOJ_1003.py
# 피보나치 수열
# 0과 1이 각각 몇 번 출력되는지 구하는 프로그램

T = int(input())
for _ in range(T):
    N = int(input())
    a,b = 1, 0 # 0과 1이 호출된 횟수
    for i in range(N):
        # 0은 1이 호출된 횟수만큼, 1은 0과 1이 호출된 횟수만큼 출력됨
        a,b = b, a+b
    print(a,b)